### ISSUE TITLE
[Broker] Claim unused PRLs

### ISSUE THREAD

**Issue by AaronVasquez** [oysterprotocol/brokernode#236]
Before deleting the private key to receive PRLs, make sure the wallet is empty and transferred to the brokers main wallet. If we delete the private key before emptying the wallet, those PRLs are lost forever.

**Comment by rfornea** [comment]
@AaronVasquez   I'm not entirely sure what this issue is referring to, could you elaborate a little more?

**Comment by AaronVasquez** [comment]
So the protocol says that the broker can keep any PRLs not buried (half if both alpha and beta work together perfectly).

So the extra PRLs that are still in the newly generated address for the session need to be transferred to the broker's main wallet? Otherwise, the broker will have a bunch of wallets from all the sessions it has processed. This wasn't described in the whitepaper, and idk if a solution was already discussed.

**Comment by rfornea** [comment]
I see what you mean.  

So, there's 2 ways we could handle it.  One way is as specified in this issue.  We go ahead and immediately send the PRLs to all those other addresses, and if we arrive at those addresses and it turns out we don't need to bury treasure there (because the other broker already did it), then we send it back to the broker's "main" address associated with that transaction.  Disadvantage is it wastes more gas, advantage is the PRL is guaranteed to be there basically immediately.  

The other way is that we wait until we have arrived at each chunk that we are supposed to attach treasure to.  If the other broker **hasn't** attached the treasure, we immediately send the PRL to that address, and attach the transaction to the tangle.  The advantage is we save on gas because we never send any PRL to an address that doesn't actually need it, which means we never have to send it back.  The disadvantage is there'd be a short delay between when the transaction is attached to the tangle versus when the PRLs are actually in the address.  But that might be okay, since the webnode has to do 1 million hashes worth of PoW before it can claim any treasure anyway.  

@brunoblock thoughts?

**Comment by rfornea** [comment]
At the moment, I am actually deleting those keys as soon as I have created the encrypted "message" field.  But we could certainly change that, and not get rid of them until we purge the session.  

I'm sort of inclined to wait until we know for sure we need the PRL, before even sending it to the address.  Just so that we don't have to do a bunch of checking in the cron jobs to see if there's balances at those addresses, and then writing logic to send the PRLs back.  

We also have remember that we have to invoke bury() on them.  I don't know if the smart contract will allow us to send the PRL to an address and then bury() later, but if we can't, then we couldn't send the PRL ahead of time because the broker wouldn't have a way to get them back, since the epoch wouldn't have passed yet.  

If we can invoke bury() on them later, then we could definitely send them back in the cron jobs.  We'd have to implement a way of checking/tracking whether it was beta broker or alpha broker that actually attached the transaction to the tangle for that particular chunk, so that the brokers would know whether to invoke bury() or whether to send the money back.

**Comment by AaronVasquez** [comment]
I might be misunderstanding things, I thought the flow was:

1. Broker is given 1 PRL to embed into a datamap
2. Let's say broker does 80% of the work, and buries 0.8 PRL
3. Once it knows it's complete, it has 0.2 PRL leftover. It would then send this to its own wallet as a reward.

**Comment by rfornea** [comment]
So there's a bunch of different "types" of wallets involved.  

1.  Treasure wallets 
These wallets are where an individual treasure for a sector is buried

2.  Transaction wallets
A broker makes a new one of these with each new transaction.  Every X amount of PRL (or every X transactions) that is earned, the broker takes the leftover PRL in all these transaction wallets and moves them to the wallet of the person who owns the broker.  Bruno said the wallet owner could configure this number, but the default could be something like 500.  So we'd need to store the addresses and private keys of these wallets.  

3.  Broker Owner's Wallet
The wallet of the person who owns the broker.  

???? Broker's Gas Wallet??
I'm not sure if this is separate from the other wallets or if it would just use the broker owner's wallet, but there needs to be a wallet that the broker can use for gas for when the webnode claims treasure.  

???? Treasure Fee Wallet??
I'm not sure if this is separate from the other wallets or if it would just use the broker owner's wallet, but there needs to be a wallet that the broker's treasure claiming fees go to.  We should probably just use the broker owner's wallet for this one.  



I was under the impression that this issue was referring to type #1 wallets, the treasure wallets.  If you were referring to type #2 wallets, then yeah I don't see any problem with just going ahead and transferring them to type #3 wallet as soon as the upload is done.  Whether we do it immediately after the upload or wait until a certain number of transactions doesn't really matter IMO.

**Comment by AaronVasquez** [comment]
Ah, I see the confusion. Sorry about not being clear.

Yes, I was referring to 2 => 3 transfer. Which I assume would be done in some cron job. Let's do the simplest solution for this deadline.

As for paying gas and fees, I was assuming owner wallet there too.

**Comment by rfornea** [comment]
Okay.  Yeah so we'll just need to make sure we don't purge the completed session until we've made the transfer from the ETH address used for the transaction.  Or the purge_completed_session task itself could be the one that kicks off the PRL transfer.  

>As for paying gas and fees, I was assuming owner wallet there too.  

I agree.

**Comment by rfornea** [comment]
The PRL transfer could take several minutes, and I don't want to have the purge_completed_session task be forced to wait around several minutes to see if the ETH transfer was successful, before it can do its job.  

So, the way I'm planning on approaching this is adding a new table called something like "completed_uploads" or something.  It would have the ETH address and key from the upload_sessions table, and a couple other properties, along with an enum indicating whether we have already attempted to claim the leftover PRLs from that address.  Within the "purge_completed_sessions" task, we'll created an entry in the "completed_uploads" table and go ahead and purge the session from "upload_sessions".    

Then I'll write another cron task that will be called something like "claim_unused_prls" or something and this task will run maybe every 5-10 minutes.  

The first thing this cron task will do is check any entries in "completed_uploads" for which we have already attempted to send the PRLs to see if the attempt was successful.  If it was successful, we delete that entry.  If it wasn't successful we'll try again.  

The second thing this cron task will do is initiate the PRL transfer for any entry that we have not yet attempted the PRL transfer, and then change the enum to indicate that we have made an attempt.


^ All of this could also just be handled in purge_completed_sessions if we wanted to, but I kinda like the idea of keeping it separate so each task isn't trying to do too much.  Also, when we eventually implement the PRL transfers the way Bruno wants, we'll want to have some separate table so we can run the task every 500 transactions or so.  We can run the task a lot less frequently and add a check to only execute it if the # of rows is greater than 500 or something, or whatever the default number is.
